### ISSUE TITLE
Extension load error "cannot find type Def in typeInfoMap"

### ISSUE THREAD

**Issue by Wolfsblvt** [zzzz465/rwxml-language-server#119]
Hey,  
no matter what I'm trying, I can't seem to get your installation to work.

I already moved Rimworld to the default install path to prevent problems that I've read about here, I entered all paths in the extension settings manually.

Now when my newly created nearly empty mod folder, the extension doesn't load up. Extension Output log shows:

```
current log level: debug
hello world! initializing @rwxml-language-server/language-server ...
initialization completed!
[warn] [f] [default] cannot find type Def in typeInfoMap. something isn't right.
[info] [r] [default] [15982c8c-ac06-4486-a584-f4d8be084e59] reloading project... reason: project initialize
[info] [r] [default] [15982c8c-ac06-4486-a584-f4d8be084e59] loading project resources...
[info] [r] [default] [15982c8c-ac06-4486-a584-f4d8be084e59] clear project...
[info] [r] [default] [15982c8c-ac06-4486-a584-f4d8be084e59] project cleared.
[info] [r] [default] [15982c8c-ac06-4486-a584-f4d8be084e59] project evaluated.
```

The extension is enabled, but doesn't show any runtime info, like other extension would do when they are loaded:
![image](https://user-images.githubusercontent.com/9962104/170563239-25248d29-0587-4605-920a-3f93fc186499.png)

This is a pretty fresh install of Visual Code.

I love the idea of your extensions, and would be really happy if I could get this to work.

**EDIT**: I also tried opening the Rimworld folder itself, and some workshop mod folders, always the same errors.
I tried to dive into the code to see which def it tried to load, but it got a bit confusing. Are you loading DLLs manually there to register type definitions?

**Comment by zzzz465** [comment]
> EDIT: I also tried opening the Rimworld folder itself, and some workshop mod folders, always the same errors.
I tried to dive into the code to see which def it tried to load, but it got a bit confusing. Are you loading DLLs manually there to register type definitions?

yes. the extension loads all required metadatas (defs, comps, etc...) from reading DLL. this is because `Ludeon Studios` doesn't allow third party distribution of the game file.

> Hey,
> no matter what I'm trying, I can't seem to get your installation to work.
> I already moved Rimworld to the default install path to prevent problems that I've read about here, I entered all paths in the extension settings manually.

I should've printed logs with more detail for easy debugging. I'll add more of it so it helps finding what is not working properly.
while I'm working, if you can set-up project environment, you can launch as VSCode debug mode to see what's going on in (extension) client. the log you're seeing is from server side.

**Comment by Wolfsblvt** [comment]
> if you can set-up project environment, you can launch as VSCode debug mode to see what's going on in (extension) client.

How do I launch VSCode in debug mode? You mean debugging the extension, so cloning your repo and trying to debug it?

Thanks for responding so quick and trying to help out!

**Comment by zzzz465** [comment]
> How do I launch VSCode in debug mode? You mean debugging the extension, so cloning your repo and trying to debug it?

you can clone the repo, and launch vscode to enter extension debug mode.

I described (how to set-up a project on readme)[https://github.com/zzzz465/rwxml-language-server#setup-devlopment-environment].

**Comment by zzzz465** [comment]
I released v0.32.0 which improve logging on client/server side.
please upload your log with the newer version so I can debug it.

**Comment by Wolfsblvt** [comment]
Trying to set up the dev environment on windows. Got it working mostly, I can build the source with `pnpm install` without errors now, but I am getting error on running with the `webpack`.  
Do I have to install that too, isn't it bundled with your files?

![image](https://user-images.githubusercontent.com/9962104/170712645-b31a01cf-68fc-4c82-8f77-b042bdc46f12.png)

So if I can't test it locally, I can't provide you with a new log yet, as I guess you haven't updated the marketplace version yet.

**Comment by Wolfsblvt** [comment]
So I guess the `run watch` is needed to start/debug the extension? When I try to run the client via VSCode it shows me this error in the client, the extension is not loaded:

![image](https://user-images.githubusercontent.com/9962104/170716561-597b2019-e407-409c-9dec-eb1da8898a49.png)

And do I have to run in the main folder, or the extension folder? Running in the extension folder gives me a different error, with an invalid config file:

![image](https://user-images.githubusercontent.com/9962104/170716780-8af07f83-7869-4bad-9200-50c425c6f744.png)

I'm so sorry for causing this much trouble.

**Comment by Wolfsblvt** [comment]
Okay, I was being a bit too fast, I removed the invalid comment from `tsconfig.json` and can now successfully run `pnpm run watch` in the vsc-extension folder.

Getting a different error now when I launch the client...

![image](https://user-images.githubusercontent.com/9962104/170717931-2e2f9e21-2b91-4600-a960-403119aafddf.png)
![image](https://user-images.githubusercontent.com/9962104/170718023-6d887528-1d81-4b42-b62e-3e4c8f23e9cb.png)

One step after the other, right..?

**Comment by Wolfsblvt** [comment]
Man, I am really diving deep right now. Hope you are still on board with me :D

I fixed some config errors and even made a pull request for those, if you wanna check that. Now I can successfully run the watch command on the root folder.

Still getting the **DependencyResourceProvider** exception:
![image](https://user-images.githubusercontent.com/9962104/170721831-013e7271-5f9f-40e5-a64f-4c988ee6dbc2.png)

```
Exception has occurred: Error: TypeInfo not known for "DependencyResourceProvider"
t InternalDependencyContainer.construct (d:\code\rwxml-language-server\vsc-extension\dist\index.js:64499:23)
	at InternalDependencyContainer.resolveRegistration (d:\code\rwxml-language-server\vsc-extension\dist\index.js:64358:24)
	at d:\code\rwxml-language-server\vsc-extension\dist\index.js:64381:30
	at Array.map (<anonymous>)
	at InternalDependencyContainer.resolveAll (d:\code\rwxml-language-server\vsc-extension\dist\index.js:64380:42)
	at activate (d:\code\rwxml-language-server\vsc-extension\dist\index.js:85054:77)
	at processTicksAndRejections (node:internal/process/task_queues:96:5)
	at async _activate (c:\Users\Wolfsblvt\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:83:8180)
	at async _waitForDepsThenActivate (c:\Users\Wolfsblvt\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:83:8122)
	at async _initialize (c:\Users\Wolfsblvt\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:83:7486)
```

The exception is happening right here: https://github.com/zzzz465/rwxml-language-server/blob/d892fe9b510656720a8245c2477eae8050d4a478/vsc-extension/src/index.ts#L61

**Comment by zzzz465** [comment]
> Okay, I was being a bit too fast, I removed the invalid comment from tsconfig.json and can now successfully run pnpm run watch in the vsc-extension folder.
https://github.com/zzzz465/rwxml-language-server/pull/122

> Still getting the DependencyResourceProvider exception:
I don't know why it is broken. I'm trying to figure it out...

**Comment by zzzz465** [comment]
@Wolfsblvt thanks for investigating issues. I merged a PR to fix the detailed issue you posted. check #123 

after a [CI build](https://github.com/zzzz465/rwxml-language-server/runs/6628466712?check_suite_focus=true) completes, the new version will be released. you can use that (takes up to ~ 30min for upload to marketplace)

**Comment by Wolfsblvt** [comment]
Thanks, I will test this again once the CI build is working again!

**Comment by zzzz465** [comment]
https://github.com/zzzz465/rwxml-language-server/actions/runs/2399557044
CI fixed.

**Comment by zzzz465** [comment]
v0.32.4 is released and now you can read client / server log from output channels.
![image](https://user-images.githubusercontent.com/46273764/170805611-4693b822-79df-4b9a-a50b-f0f3243ad7b6.png)

**Comment by Wolfsblvt** [comment]
Nice, that's a pretty detailed log now!

So, either I am not understanding correctly what the Extension actually is doing, or it doesn't really load up any of the features.  
I feel like it should let me F12 jump to parents in ThingDef ParentName definitions. Which kind of additional syntax highlighting  and auto completion is included, so that I can test that?

I also tried opening several mods in my workshop folder, and none really showed any features available either, with similar logs. Nothing happening on opening the "Data" folder in Rimworld install either.

So am I understanding something wrong, or is it not working for me?

Attached both client and server logs. Server still has the one with "Def" not found that looks like it might be culprit.

**Server Log**
```
log level changed to debug
[info]	[undefined]:	hello world! initializing @rwxml-language-server/server ...
[info]	[undefined]:	initialization completed!
(node:52932) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(Use `Code --trace-warnings ...` to show where the warning was created)
[info]	[About]:	about file changed.
[debug]	[About]:	updating paths because about.xml path is changed.
[debug]	[About]:	about.xml path: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/About/About.xml
[debug]	[About]:	rootDirectory path: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch
[debug]	[LoadFolder]:	reloading because About.xml path is changed. new source: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/LoadFolders.xml
[debug]	[LoadFolder]:	updated workspaces: [
  {
    "relativeDirectories": [
      "."
    ],
    "rootDir": {
      "$mid": 1,
      "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
      "scheme": "file"
    },
    "version": "default"
  },
  {
    "relativeDirectories": [
      "."
    ],
    "rootDir": {
      "$mid": 1,
      "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
      "scheme": "file"
    },
    "version": "1.3"
  }
]
[debug]	[LoadFolder]:	default workspace: {
  "relativeDirectories": [
    "."
  ],
  "rootDir": {
    "$mid": 1,
    "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
    "scheme": "file"
  },
  "version": "default"
}
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[info]	[ProjectManager]:	supportedVersions added: [
  "default",
  "1.3"
]
[info]	[ProjectManager]:	supportedVersions deleted: []
[debug]	[AboutMetadata]:	reloading because About.xml path is changed. new source: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/About/metadata_rwxml.xml
[debug]	[AboutMetadata]:	aboutMetadata updated.
[debug]	[AboutMetadata]:	items: {}
[debug]	[AboutMetadata]:	default: undefined
[debug]	[ModDependencyBags]:	updating state because AboutMetadata is changed.
[debug]	[ModDependencyBags]:	updating state because about.xml is changed.
[info]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	reloading project... reason: project initialize
[info]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	loading project resources...
[info]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	clear project...
[debug]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	requesting typeInfo. count: 0
[debug]	[TypeInfoMapProvider]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	received typeInfo from client, length: 15063
[info]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	project cleared.
[info]	[Project]	(c0ac73c5-d602-4c13-ae49-9ce7fb1fe744):	project evaluated.
[info]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	reloading project... reason: dll is changed.
[info]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	loading project resources...
[info]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	clear project...
[debug]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	current project file dll count: 1
[debug]	[TypeInfoMapProvider]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	requesting typeInfo. count: 1
[debug]	[TypeInfoMapProvider]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	received typeInfo from client, length: 15102
[info]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	project cleared.
[info]	[Project]	(ade22d38-9ebf-4faa-be6d-80051a0e869d):	project evaluated.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[info]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	reloading project... reason: project initialize
[info]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	loading project resources...
[info]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	clear project...
[debug]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(516df1c7-6651-4c82-b978-4245ec575ee9):	requesting typeInfo. count: 0
[info]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	reloading project... reason: project initialize
[info]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	loading project resources...
[info]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	clear project...
[debug]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	requesting typeInfo. count: 0
[info]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	reloading project... reason: project initialize
[info]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	loading project resources...
[info]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	clear project...
[debug]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	requesting typeInfo. count: 0
[debug]	[TypeInfoMapProvider]	(516df1c7-6651-4c82-b978-4245ec575ee9):	received typeInfo from client, length: 15063
[info]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	project cleared.
[info]	[Project]	(516df1c7-6651-4c82-b978-4245ec575ee9):	project evaluated.
[debug]	[TypeInfoMapProvider]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	received typeInfo from client, length: 15063
[info]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	project cleared.
[info]	[Project]	(0567ae92-bef2-4aeb-ad2b-98479baa0a36):	project evaluated.
[debug]	[TypeInfoMapProvider]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	received typeInfo from client, length: 15063
[info]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	project cleared.
[info]	[Project]	(35a83c59-a972-4502-8a26-9aca9d66dd6a):	project evaluated.
```

**Client Log**
```
[info]	[undefined]:	registering commands...
[info]	[undefined]:	initializing Language Server...
[debug]	[undefined]:	server module absolute path: c:\Users\Wolfsblvt\.vscode\extensions\madeline.rwxml-lang-serv-0.32.4\dist\language-server\index.js
[info]	[undefined]:	initializing resourceProviders...
[debug]	[ProjectWatcher]:	ModManager watching directories: [
  "c:\\Program Files (x86)\\Steam\\steamapps\\common\\RimWorld\\Data",
  "c:\\Program Files (x86)\\Steam\\steamapps\\common\\RimWorld\\Mods",
  "c:\\Program Files (x86)\\Steam\\steamapps\\workshop\\content\\294100"
]
[info]	[undefined]:	checking Runtime TypeInfo Extractor available...
[info]	[undefined]:	register lsp features...
[info]	[undefined]:	initialize Project Watcher...
[info]	[undefined]:	initialization completed.
[debug]	[ProjectWatcher]:	sending initial load files. count: 5
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Mlie.IndustrialRollers, version: default
[debug]	[DependencyResourceProvider]:	found 43 dependency files in packageId: Mlie.IndustrialRollers
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Ludeon.RimWorld, version: default
[debug]	[DependencyResourceProvider]:	found 492 dependency files in packageId: Ludeon.RimWorld
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Mlie.IndustrialRollers, version: 1.3
[debug]	[DependencyResourceProvider]:	found 44 dependency files in packageId: Mlie.IndustrialRollers
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Ludeon.RimWorld, version: 1.3
[debug]	[DependencyResourceProvider]:	found 492 dependency files in packageId: Ludeon.RimWorld
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 1
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
```

**Comment by Wolfsblvt** [comment]
Okay, I tested a bit more, digged through some old issues here, and found another possible cause. I got a few mods that got some features loaded, and I noticed they all had a `LoadFolders.xml` file. I added one to my test mod, and it started to load up.  
Is this file needed for this extension to work?  
I can live with that, but it would make looking into other mods with this extension a bit more difficult.

The `cannot find type Def in typeInfoMap. something isn't right.` error is still there though, and I don't know if that might cause issues?  
I made a few test Defs, just copy&paste, and most of the things I expected seem to be working. Some issues though.
- The ParentName tag does have a popup of the parent definition, but F12/Go to definition does not work there. Works in other popup/underlined places, like inside researchPrerequisites.
- My own abstract def has no quick info line above. Should show that it is referenced once
- Go to references on the abstract def doesn't work

Something wrong on my end? Are those known issues? Should I open a new GitHub issue?

![image](https://user-images.githubusercontent.com/9962104/170808840-99720c6a-ad1c-46db-a2ff-cadd2e6d0345.png)

**Server Log**
```
log level changed to debug
[info]	[undefined]:	hello world! initializing @rwxml-language-server/server ...
[info]	[undefined]:	initialization completed!
(node:51232) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(Use `Code --trace-warnings ...` to show where the warning was created)
[info]	[About]:	about file changed.
[debug]	[About]:	updating paths because about.xml path is changed.
[debug]	[About]:	about.xml path: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/About/About.xml
[debug]	[About]:	rootDirectory path: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch
[debug]	[LoadFolder]:	reloading because About.xml path is changed. new source: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/LoadFolders.xml
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[info]	[ProjectManager]:	supportedVersions added: [
  "default",
  "1.3"
]
[info]	[ProjectManager]:	supportedVersions deleted: []
[debug]	[AboutMetadata]:	reloading because About.xml path is changed. new source: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/About/metadata_rwxml.xml
[debug]	[AboutMetadata]:	aboutMetadata updated.
[debug]	[AboutMetadata]:	items: {}
[debug]	[AboutMetadata]:	default: undefined
[debug]	[ModDependencyBags]:	updating state because AboutMetadata is changed.
[debug]	[ModDependencyBags]:	updating state because about.xml is changed.
[debug]	[LoadFolder]:	updated workspaces: [
  {
    "relativeDirectories": [
      "/"
    ],
    "rootDir": {
      "$mid": 1,
      "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
      "scheme": "file"
    },
    "version": "1.3"
  },
  {
    "relativeDirectories": [
      "."
    ],
    "rootDir": {
      "$mid": 1,
      "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
      "scheme": "file"
    },
    "version": "default"
  }
]
[debug]	[LoadFolder]:	default workspace: {
  "relativeDirectories": [
    "."
  ],
  "rootDir": {
    "$mid": 1,
    "path": "/c:/Program Files (x86)/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch",
    "scheme": "file"
  },
  "version": "default"
}
[info]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	reloading project... reason: project initialize
[info]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	loading project resources...
[info]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	clear project...
[debug]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	requesting typeInfo. count: 0
[debug]	[TypeInfoMapProvider]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	received typeInfo from client, length: 15063
[info]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	project cleared.
[info]	[Project]	(64f3a2d5-40a3-4ea3-879b-e4f156b62e9f):	project evaluated.
[info]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	reloading project... reason: dll is changed.
[info]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	loading project resources...
[info]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	clear project...
[debug]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	current project file dll count: 1
[debug]	[TypeInfoMapProvider]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	requesting typeInfo. count: 1
[debug]	[DiagnosticsProvider]:	[default] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ThingDefTest.xml, items: 0
[debug]	[DiagnosticsProvider]:	[default] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ResearchProjectDef/Research.xml, items: 0
[debug]	[TypeInfoMapProvider]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	received typeInfo from client, length: 15102
[info]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	project cleared.
[info]	[Project]	(f3e678e8-3626-45ff-82d7-3f21a254113f):	project evaluated.
[debug]	[DiagnosticsProvider]:	[1.3] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ThingDefTest.xml, items: 0
[debug]	[DiagnosticsProvider]:	[1.3] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ResearchProjectDef/Research.xml, items: 0
[debug]	[DiagnosticsProvider]:	[default] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ResearchProjectDef/Research.xml, items: 0
[debug]	[DiagnosticsProvider]:	[1.3] send diagnostics to uri: file:///c%3A/Program%20Files%20%28x86%29/Steam/steamapps/common/RimWorld/Mods/IndustrialRollersResearch/Defs/ResearchProjectDef/Research.xml, items: 0
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[warn]	[DefManager]:	cannot find type Def in typeInfoMap. something isn't right.
[info]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	reloading project... reason: project initialize
[info]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	loading project resources...
[info]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	clear project...
[debug]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	requesting typeInfo. count: 0
[info]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	reloading project... reason: project initialize
[info]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	loading project resources...
[info]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	clear project...
[debug]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	requesting typeInfo. count: 0
[info]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	reloading project... reason: project initialize
[info]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	loading project resources...
[info]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	clear project...
[debug]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	current project file dll count: 0
[debug]	[TypeInfoMapProvider]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	requesting typeInfo. count: 0
[debug]	[TypeInfoMapProvider]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	received typeInfo from client, length: 15063
[info]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	project cleared.
[info]	[Project]	(2a03c57e-d6da-4643-a8df-cd7115df36ae):	project evaluated.
[debug]	[TypeInfoMapProvider]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	received typeInfo from client, length: 15063
[info]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	project cleared.
[info]	[Project]	(9b62cb0e-ea0c-45dd-81fc-9ae59d72abf5):	project evaluated.
[debug]	[TypeInfoMapProvider]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	received typeInfo from client, length: 15063
[info]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	project cleared.
[info]	[Project]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	project evaluated.
```

**Client Log**
```
[info]	[undefined]:	registering commands...
[info]	[undefined]:	initializing Language Server...
[debug]	[undefined]:	server module absolute path: c:\Users\Wolfsblvt\.vscode\extensions\madeline.rwxml-lang-serv-0.32.4\dist\language-server\index.js
[info]	[undefined]:	initializing resourceProviders...
[debug]	[ProjectWatcher]:	ModManager watching directories: [
  "c:\\Program Files (x86)\\Steam\\steamapps\\common\\RimWorld\\Data",
  "c:\\Program Files (x86)\\Steam\\steamapps\\common\\RimWorld\\Mods",
  "c:\\Program Files (x86)\\Steam\\steamapps\\workshop\\content\\294100"
]
[info]	[undefined]:	checking Runtime TypeInfo Extractor available...
[info]	[undefined]:	register lsp features...
[info]	[undefined]:	initialize Project Watcher...
[info]	[undefined]:	initialization completed.
[debug]	[ProjectWatcher]:	sending initial load files. count: 6
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Mlie.IndustrialRollers, version: default
[debug]	[DependencyResourceProvider]:	found 43 dependency files in packageId: Mlie.IndustrialRollers
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Ludeon.RimWorld, version: default
[debug]	[DependencyResourceProvider]:	found 492 dependency files in packageId: Ludeon.RimWorld
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Mlie.IndustrialRollers, version: 1.3
[debug]	[DependencyResourceProvider]:	found 44 dependency files in packageId: Mlie.IndustrialRollers
[debug]	[DependencyResourceProvider]:	received dependency request for packageId: Ludeon.RimWorld, version: 1.3
[debug]	[DependencyResourceProvider]:	found 492 dependency files in packageId: Ludeon.RimWorld
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 1
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
[debug]	[CachedTypeInfoProvider]:	received typeInfo request. uris count: 0
```

**Comment by zzzz465** [comment]
> [debug]	[TypeInfoMapProvider]	(71a6da3d-8513-4612-bd4b-ab3eb8343ee8):	received typeInfo from client, length: 15063

it seems dll load works fine.  

> Okay, I tested a bit more, digged through some old issues here, and found another possible cause. I got a few mods that got some features loaded, and I noticed they all had a LoadFolders.xml file. I added one to my test mod, and it started to load up.
Is this file needed for this extension to work?
I can live with that, but it would make looking into other mods with this extension a bit more difficult.

extension can run with/without `loadfolders.xml` and it should works. if it doesn't, it need to be addressed in a new issue.

> The cannot find type Def in typeInfoMap. something isn't right. error is still there though, and I don't know if that might cause issues?

this log happens when the server receives typeInfos from client and there's no data for `<Def>` which is the base component of all others. meaning that the received data is corrupted.
it displayed at least once because typeInfo is not available yet when initializing project environment. because the data is loaded after the initialization.

> The ParentName tag does have a popup of the parent definition, but F12/Go to definition does not work there. Works in other popup/underlined places, like inside researchPrerequisites.

some nodes are not registered as linkable. it is not intended and needs to be fix.

> My own abstract def has no quick info line above. Should show that it is referenced once. Checking other mods that work as well none of the abstract defs have that line above.
> Go to references on the abstract def doesn't work

quick info is called "CodeLens". anyway, the reason why it's broken is a bit complex.
there's two refernce types, 1. `<defName>`, 2. `Name` attribute.
language server manages both types on resolving, but some features supports only one. that's why it doesn't pop up quick info.

you can see [codelens code here](https://github.com/zzzz465/rwxml-language-server/blob/7c5ac82fcd77ca3d282d06895daa780dfa057a04/language-server/src/features/codeLens.ts#L33-L53)
as you can see, it checks `<defName>` value to determine whether to display codelens or not.

it is not a intended feature and needs to be fix like above.

**Comment by zzzz465** [comment]
> Okay, I tested a bit more, digged through some old issues here, and found another possible cause. I got a few mods that got some features loaded, and I noticed they all had a LoadFolders.xml file. I added one to my test mod, and it started to load up.
Is this file needed for this extension to work?

I just remembered that this is a known issue.

**Comment by Wolfsblvt** [comment]
Cheers!

Thanks for taking your time to responding to me and my issues.

If you need help I could see if I can learn how typescript works and help out with code.  
I can create issues for the things I noticed, so we don't have this big monster issue that's hard to follow.

Or tell how I can buy you a coffee, for the headaches I caused.

**Comment by zzzz465** [comment]
I appreciate your coffee but I can get it on my own :3
github star and sharing this is enough for me.
feel free to make an issue if you find any trouble while using it.

you can close this issue if the problem is solved.

**Comment by Wolfsblvt** [comment]
Like your attitude <3 I stared already and I am going to share it whenever I can!

I guess the original problem here is solved, extension is loading and working well, I am going to close it.  
I'll open issues for the other issues I experience.
